extensions/amp-form/0.1/form-validators.js

```diff
@@ -0,0 +1,347 @@
+    for (let i = 0; i < inputs.length; i++) {
+      this.hideValidationFor(inputs[i]);
+    }
+  }
+
+  /**
+   * @param {!Element} input
+   * @param {!string} invalidType
+   * @return {?Element}
+   */
+  getValidationFor(input, invalidType) {
+    if (!input.id) {
+      return null;
+    }
+
+    return this.doc.querySelector(
+        `[visible-when-invalid=${invalidType}][validation-for=${input.id}]`);
+  }
+
+  /**
+   * @param {!Element} input
+   * @param {string} invalidType
+   */
+  showValidationFor(input, invalidType) {
+    const validation = this.getValidationFor(input, invalidType);
+    if (!validation) {
+      return;
+    }
+
+    if (!validation.textContent.trim()) {
+      validation.textContent = input.validationMessage;
+    }
+
+    input.setAttribute('aria-invalid', 'true');
+    validation.classList.add('visible');
+  }
+
+  /**
+   * @param {!Element} input
+   * @return {boolean} Whether a validation was visible and was successfully hidden.
+   */
+  hideValidationFor(input) {
+    const visibleValidation = this.getVisibleValidationFor(input);
+    if (!visibleValidation) {
+      return false;
+    }
+
+    input.removeAttribute('aria-invalid');
+    visibleValidation.classList.remove('visible');
+    return true;
+  }
+
+  /**
+   * @param {!Element} input
+   * @return {?Element}
+   */
+  getVisibleValidationFor(input) {
+    if (!input.id) {
+      return null;
+    }
+    return this.doc.querySelector(
+        `[visible-when-invalid][validation-for=${input.id}].visible`);;
+  }
+
+  /**
+   * Whether an input should validate after an interaction.
+   * @param {!Element} input
+   * @return {boolean}
+   */
+  shouldValidateOnInteraction(input) {
+    return !!this.getVisibleValidationFor(input);
+  }
+
+  /**
+   * @param {!Event} event
+   */
+  onInteraction(event) {
+    const input = /** @type {!Element} */ (event.target);
+    const shouldValidate = this.shouldValidateOnInteraction(input);
+    this.hideValidationFor(input);
+    if (shouldValidate && !event.target.checkValidity()) {
+      this.reportInput(input);
+    }
+  }
+
+  /** @override */
+  onBlur(event) {
+    this.onInteraction(event);
+  }
+
+  /** @override */
+  onInput(event) {
+    this.onInteraction(event);
+  }
+}
+
+
+class ShowFirstOnSubmitValidator extends AbstractCustomValidator {
+
+  /** @override */
+  report() {
+    this.hideAllValidations();
+    const inputs = this.form.querySelectorAll('input,select,textarea');
+    for (let i = 0; i < inputs.length; i++) {
+      if (!inputs[i].checkValidity()) {
+        this.reportInput(inputs[i]);
+        inputs[i]./*REVIEW*/focus();
+        break;
+      }
+    }
+  }
+
+}
+
+
+class ShowAllOnSubmitValidator extends AbstractCustomValidator {
+
+  /** @override */
+  report() {
+    this.hideAllValidations();
+    let firstInvalidInput = null;
+    const inputs = this.form.querySelectorAll('input,select,textarea');
+    for (let i = 0; i < inputs.length; i++) {
```

Comment:
This might be cleaner as two loops:

```js
for (let i = 0 ... inputs.length) {
  if (!valid) {
    reportInput(input);
    input.focus();
    break;
  }
}

for (i ... inputs.length) {
  if (!valid) {
    reportInput(input);
  }
}